Eliminate memory allocations in MVCCComputeStats.

petermattis opened the pull request:
Reworked the MVCCComputeStats benchmark to use 64MB ranges. Added
Iterator.unsafe{Key,Value} to avoid copying data. Added mvccDecodeKey to
avoid allocating memory during key decoding.

```
name                                old time/op    new time/op     delta
MVCCComputeStats1Version8Bytes-8       4.05s ± 2%      2.09s ± 1%   -48.39%  (p=0.000 n=9+9)
MVCCComputeStats1Version32Bytes-8      2.85s ± 2%      1.47s ± 1%   -48.37%  (p=0.000 n=8+10)
MVCCComputeStats1Version256Bytes-8     796ms ± 1%      404ms ± 1%   -49.25%  (p=0.000 n=10+10)

name                                old speed      new speed       delta
MVCCComputeStats1Version8Bytes-8    16.6MB/s ± 1%   32.1MB/s ± 1%   +93.76%  (p=0.000 n=9+9)
MVCCComputeStats1Version32Bytes-8   23.4MB/s ± 7%   45.7MB/s ± 1%   +95.25%  (p=0.000 n=9+10)
MVCCComputeStats1Version256Bytes-8  84.3MB/s ± 1%  166.1MB/s ± 1%   +97.04%  (p=0.000 n=10+10)
```

The remaining perf issue is the overhead of individual cgo calls. Moving stats computation into C++ would give a pretty massive speedup, but would involve duplicating some of the stats computation code. 

I got rid of the multi-version benchmarks due to laziness about computing the per key/value overhead.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3096)

<!-- Reviewable:end -->

petermattis committed:
Eliminate memory allocations in MVCCComputeStats.

Reworked the MVCCComputeStats benchmark to use 64MB ranges. Added
Iterator.unsafe{Key,Value} to avoid copying data. Added mvccDecodeKey to
avoid allocating memory during key decoding.

name                                old speed      new speed       delta
MVCCComputeStats1Version8Bytes-8    16.6MB/s ± 1%   32.1MB/s ± 1%   +93.76%  (p=0.000 n=9+9)
MVCCComputeStats1Version32Bytes-8   23.4MB/s ± 7%   45.7MB/s ± 1%   +95.25%  (p=0.000 n=9+10)
MVCCComputeStats1Version256Bytes-8  84.3MB/s ± 1%  166.1MB/s ± 1%   +97.04%  (p=0.000 n=10+10)

petermattis commented:
Oh, guess I should run the tests. Don't look at this yet.

tamird commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

<sup>**[storage/replica_data_iter.go, line 98 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096#-K2nNYRZc2OqYsBNY3QG)** ([raw file](https://github.com/cockroachdb/cockroach/blob/be6341e2add363fee69d87db73a079f29f3b4359/storage/replica_data_iter.go#L98)):</sup>
any reason to specifically call these methods on the field? it's embedded now.
<!-- Reviewable comment -K2nNYR_pegn8gxqSh6j:-285422839 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096)
<!-- Sent from Reviewable.io -->

petermattis commented:
Ok, this is ready for a look. 

I have a follow-up change which adjusts the C <-> Go interface to rocksdb and reduces the number of cgo calls when using an iterator by 4x.

petermattis commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

<sup>**[storage/replica_data_iter.go, line 98 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096#-K2nNYRZc2OqYsBNY3QG)** ([raw file](https://github.com/cockroachdb/cockroach/blob/be6341e2add363fee69d87db73a079f29f3b4359/storage/replica_data_iter.go#L98)):</sup>
Search&replace was the reason. Will fix.
<!-- Reviewable comment -K2oELIAwqc4H3CwpPG1:-956870138 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096)
<!-- Sent from Reviewable.io -->

petermattis committed:
Reduce cgo crossing during rocksdb iteration.

Changed the various DBIter{Seek,Next,Prev} methods to return a
DBIterState struct that contains whether the iterator is valid and, if
it is valid, the current key and value. Removed
DBIter{Valid,Key,Value}(). Cache DBIterState in rocksDBIterator and use
its info to implement the {Valid,Key,Value}() methods. Using DBIterState
instead of the DBIter{Valid,Key,Value}() functions reduces the number of
cgo calls per loop iteration from 4 to 1.

name                                old speed      new speed       delta
MVCCScan1Version1Row-8              94.9MB/s ± 4%  100.3MB/s ± 4%   +5.68%         (p=0.000 n=8+10)
MVCCScan1Version10Rows-8             146MB/s ±10%    156MB/s ± 6%   +6.23%        (p=0.009 n=10+10)
MVCCScan1Version100Rows-8            173MB/s ± 6%    189MB/s ± 4%   +9.39%        (p=0.000 n=10+10)
MVCCScan1Version1000Rows-8           186MB/s ±10%    194MB/s ±11%   +4.61%        (p=0.043 n=10+10)
MVCCScan10Versions1Row-8             101MB/s ±11%    100MB/s ±12%     ~           (p=0.896 n=10+10)
MVCCScan10Versions10Rows-8           322MB/s ± 7%    343MB/s ± 7%   +6.37%        (p=0.029 n=10+10)
MVCCScan10Versions100Rows-8          355MB/s ± 9%    360MB/s ±11%     ~           (p=0.796 n=10+10)
MVCCScan10Versions1000Rows-8         345MB/s ± 9%    366MB/s ± 4%   +6.21%          (p=0.003 n=9+9)
MVCCGet1Version-8                    110MB/s ± 3%    112MB/s ± 1%   +1.76%        (p=0.009 n=10+10)
MVCCGet10Versions-8                  157MB/s ± 8%    161MB/s ± 8%     ~           (p=0.225 n=10+10)
MVCCComputeStats1Version8Bytes-8    30.4MB/s ± 1%   60.6MB/s ± 1%  +99.61%        (p=0.000 n=10+10)
MVCCComputeStats1Version32Bytes-8   43.1MB/s ± 2%   85.6MB/s ± 1%  +98.92%        (p=0.000 n=10+10)
MVCCComputeStats1Version256Bytes-8   153MB/s ± 2%    299MB/s ± 2%  +95.70%         (p=0.000 n=9+10)

petermattis commented:
Added the cgo commit to this PR which gives a nice little across the board speedup to `MVCC` operations and a massive speedup to `MVCCComputeStats`. PTAL.

vivekmenezes commented on storage/engine/rocksdb/db.h:
```diff
@@ -38,6 +38,12 @@ typedef struct {
 } DBString;
 
+typedef struct {
+  int valid;
```
bool?

vivekmenezes commented on storage/replica_data_iter.go:
```diff
@@ -75,62 +74,36 @@ func newReplicaDataIterator(d *roachpb.RangeDescriptor, e engine.Engine) *replic
 func (ri *replicaDataIterator) Close() {
 	ri.curIndex = len(ri.ranges)
-	ri.iter.Close()
+	ri.Iterator.Close()
```
ri.Close() would work right?

vivekmenezes commented on storage/engine/mvcc.go:
```diff
@@ -1591,21 +1591,25 @@ func MVCCComputeStats(iter Iterator, nowNanos int64) (MVCCStats, error) {
 	ms := MVCCStats{LastUpdateNanos: nowNanos}
 	first := false
 	meta := &MVCCMetadata{}
+	scratch := make([]byte, 0, 128)
```
Is there value to allocating an array on the stack and building a slice out of that?

tschottdorf commented:
Nice. At 60mb/s we're still going to sit around for a second, though. If we need to be faster in practice I think we should be able to compute the base stats asynchronously before the split (and keep track of the delta until the actual split trigger), but it's going to be a little involved.
LGTM
<!-- Reviewable comment -K2rAf5_OeykBos3yzzY:-600498825 -->

---

Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

<sup>**[storage/engine/engine.go, line 61 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096#-K2r89QqHLyXup4Rv7LC)** ([raw file](https://github.com/cockroachdb/cockroach/blob/01aa15a35b08d6d6a70172e1436b27ff16eefc79/storage/engine/engine.go#L61)):</sup>
add comments about the way in which these are unsafe (my understanding is that the expose memory managed by RocksDB, so presumably holding on to these across an iteration is forbidden)?
<!-- Reviewable comment -K2r89QqHLyXup4Rv7LD:2038631829 -->

---

<sup>**[storage/engine/rocksdb_test.go, line 488 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096#-K2r9wM30QxKLh3M69fb)** ([raw file](https://github.com/cockroachdb/cockroach/blob/01aa15a35b08d6d6a70172e1436b27ff16eefc79/storage/engine/rocksdb_test.go#L488)):</sup>
not sure what to make of this comment.
<!-- Reviewable comment -K2r9wM30QxKLh3M69fc:387926840 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096)
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

<sup>**[storage/engine/rocksdb_test.go, line 488 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096#-K2r9wM30QxKLh3M69fb)** ([raw file](https://github.com/cockroachdb/cockroach/blob/01aa15a35b08d6d6a70172e1436b27ff16eefc79/storage/engine/rocksdb_test.go#L488)):</sup>
Ha, that one is my doing.
<!-- Reviewable comment -K2rKmVKQL6ziqtQMiVz:1877526102 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096)
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM
<!-- Reviewable comment -K2rnU4bX5WKWuVIdaDc:676834233 -->

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

<sup>**[storage/replica_data_iter.go, line 41 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096#-K2rmAhSB0jecpY1RfIi)** ([raw file](https://github.com/cockroachdb/cockroach/blob/01aa15a35b08d6d6a70172e1436b27ff16eefc79/storage/replica_data_iter.go#L41)):</sup>
Why this change? Does it reduce allocations somehow?
<!-- Reviewable comment -K2rmAhSB0jecpY1RfIj:742334630 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096)
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

<sup>**[storage/replica_data_iter.go, line 41 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096#-K2rmAhSB0jecpY1RfIi)** ([raw file](https://github.com/cockroachdb/cockroach/blob/01aa15a35b08d6d6a70172e1436b27ff16eefc79/storage/replica_data_iter.go#L41)):</sup>
it avoids defining passthrough methods on `replicaDataIterator`, see the deleted code below.
<!-- Reviewable comment -K2roX_x_sdz7pMj5Slk:-636842982 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096)
<!-- Sent from Reviewable.io -->

petermattis commented:
@tschottdorf We can also implement your suggestion to only compute the stats for one of the new ranges and do some arithmetic to compute the stats for the other half.
<!-- Reviewable comment -K2s1nv-MEQOaQYRvxzq:-450739298 -->

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

<sup>**[storage/engine/engine.go, line 61 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096#-K2r89QqHLyXup4Rv7LC)** ([raw file](https://github.com/cockroachdb/cockroach/blob/01aa15a35b08d6d6a70172e1436b27ff16eefc79/storage/engine/engine.go#L61)):</sup>
Yes, these look like go types, but the memory is managed by C. The returned values will be invalid when the iterator is destroyed or upon the next call to `{Next,Seek,etc}`. I'll add a comment.
<!-- Reviewable comment -K2s241DDKsPuwWJzDTv:1902601002 -->

---

<sup>**[storage/engine/mvcc.go, line 1594 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096#-K2nJ6aLKhyhknHnfqfI-r2-1594)** ([raw file](https://github.com/cockroachdb/cockroach/blob/01aa15a35b08d6d6a70172e1436b27ff16eefc79/storage/engine/mvcc.go#L1594)):</sup>
You can't allocate an array on the stack. You can declare an array, but the compiler would determine the memory escapes and allocate it on the heap when it notices a slice is created from the array.
<!-- Reviewable comment -K2s2Ywp_z0fDiwHbwYx:-1428072758 -->

---

<sup>**[storage/engine/rocksdb/db.h, line 42 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096#-K2q2Ae0O1wWMKIs6x_S-r2-42)** ([raw file](https://github.com/cockroachdb/cockroach/blob/01aa15a35b08d6d6a70172e1436b27ff16eefc79/storage/engine/rocksdb/db.h#L42)):</sup>
Yeah, but `bool` is a C99-ism that I've had difficulty using with cgo in the past.
<!-- Reviewable comment -K2s2qBq_b0IKH0wv7a0:1379498380 -->

---

<sup>**[storage/engine/rocksdb_test.go, line 488 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096#-K2r9wM30QxKLh3M69fb)** ([raw file](https://github.com/cockroachdb/cockroach/blob/01aa15a35b08d6d6a70172e1436b27ff16eefc79/storage/engine/rocksdb_test.go#L488)):</sup>
I'll fix.
<!-- Reviewable comment -K2s303SGomCKe9G_UHa:-236484810 -->

---

<sup>**[storage/replica_data_iter.go, line 41 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096#-K2rmAhSB0jecpY1RfIi)** ([raw file](https://github.com/cockroachdb/cockroach/blob/01aa15a35b08d6d6a70172e1436b27ff16eefc79/storage/replica_data_iter.go#L41)):</sup>
More importantly, we can't do pass-through versions of `unsafe{Key,Value}` because those methods are not exported. There is no performance change from embedding for using a field.
<!-- Reviewable comment -K2s31w_ZkwPwTTtLtt2:-2145994961 -->

---

<sup>**[storage/replica_data_iter.go, line 77 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096#-K2nK0VW1OTRZOrcZ_Fx-r2-77)** ([raw file](https://github.com/cockroachdb/cockroach/blob/01aa15a35b08d6d6a70172e1436b27ff16eefc79/storage/replica_data_iter.go#L77)):</sup>
That would call `replicataDataIterator.Close` recursively, not what we want to do here.
<!-- Reviewable comment -K2s3GbAhKVbx4Yixph8:-948626742 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096)
<!-- Sent from Reviewable.io -->

tschottdorf commented:
yes, and we should definitely do that. I was thinking about the scenario in which even that isn't enough.
LGTM
<!-- Reviewable comment -K2s4-MiMoCxLnGnlm3u:-1971927153 -->

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3096)
<!-- Sent from Reviewable.io -->

petermattis commented:
@tschottdorf I'm pretty sure we could also get a speedup by moving all of the `MVCCComputeStats` computation into C++. Or perhaps be batching iteration with a `NextN()` that returns the next N rows. That might be easier than computing the stats asynchronously before the split.

tschottdorf commented:
sure, if that somehow gets the 32mb of data processed in a way that takes on the order of .01s, then that's the way to go.

petermattis commented:
Well, that's not gong to happen.

petermattis committed:
Update UI deps.
